velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -20,12 +20,81 @@
 #include "velox/connectors/hive/HiveConnector.h"
 #include "velox/connectors/hive/HiveWriteProtocol.h"
 #include "velox/dwio/dwrf/writer/Writer.h"
+#include "velox/exec/OperatorUtils.h"
 
 using namespace facebook::velox::dwrf;
 using WriterConfig = facebook::velox::dwrf::Config;
 
 namespace facebook::velox::connector::hive {
 
+namespace {
+std::vector<column_index_t> getPartitionChannels(
+    std::shared_ptr<const HiveInsertTableHandle> insertTableHandle) {
+  std::vector<column_index_t> channels;
+
+  for (column_index_t i = 0; i < insertTableHandle->inputColumns().size();
+       i++) {
+    if (insertTableHandle->inputColumns()[i]->isPartitionKey()) {
+      channels.push_back(i);
+    }
+  }
+
+  return channels;
+}
+
+RowTypePtr getChannelTypes(
+    RowTypePtr rowType,
+    const std::vector<column_index_t>& channels) {
+  std::vector<std::string> names;
+  std::vector<velox::TypePtr> types;
+  names.reserve(channels.size());
+  types.reserve(channels.size());
+
+  for (column_index_t index : channels) {
+    names.push_back(rowType->nameOf(index));
+    types.push_back(rowType->childAt(index));
+  }
+
+  return ROW(std::move(names), std::move(types));
+}
+
+uint64_t getMaxPartitionId(
```

Comment:
I wonder if we can avoid this computation by fetching max partition ID from the ID generator.

Reply:
We probably cannot.  I'm calling this getMaxPartitionId to get the max ID for current input, or the most recent input that is fed into the ID generator. On the other hand, I think we can only easily get the max ID from the ID generator or VectorHasher for ALL prior inputs. Max ID for current input might be smaller than max ID for all inputs globally.

Reply:
@gggrace14 Is it important that max ID for the recent input and not global?